### PR TITLE
feat: substream eth indexers and cross-chain db

### PR DESCRIPTION
**Pull request by hthieu1110**
- [x] squad eth integration 
- [x] integrate substreams indexer into dapp
- [x] implement needed handlers for eth
- [x] rewards distribution

NOTE: 
- [ ] test sending tori rewards due to change of rpc endpoint from .env => network package

### CODE REVIEW

File: .env [hunk 1]
```diff
@@ -31,3 +28,11 @@ PRICES_SERVICE_URI=prices.mainnet.teritori.com:443
 DISPLAYED_NETWORKS_IDS=teritori,teritori-testnet,ethereum,ethereum-goerli,cosmos-hub,juno,osmosis,osmosis-testnet,solana,gno-testnet,gno-dev
 SELECTABLE_NETWORKS_IDS=teritori,teritori-testnet,ethereum,ethereum-goerli,osmosis,osmosis-testnet,gno-testnet,gno-dev
 IMG_PROXY_URL=https://imgproxy.tools.teritori.com/insecure/
+
+SUBSTREAMS_API_TOKEN=eyJhbGciOiJLTVNFUzI1NiIsInR5cCI6IkpXVCJ9.eyJleHAiOjE5ODg3MDI2NTEsImp0aSI6ImU3NzhjNGZhLTk1YjItNGJmMy04NGNmLTJjZjk0NTcyZWY3ZSIsImlhdCI6MTY3MzM0MjY1MSwiaXNzIjoiZGZ1c2UuaW8iLCJzdWIiOiJ1aWQ6MGx5bGU3M2EzMzIxMThjOWQzNzg4IiwidGllciI6ImZyZWUtdjEiLCJ2IjoxLCJ1c2ciOiJzZXJ2ZXIiLCJha2kiOiI2ODFmYzg5YTBmOGRkYTUyZjg3ZWZmZTRmMzQ5NDdjNDU1NTNlODg0NGQ5MTA1ZThhZTZkMDBmM2ZkMDg0YTEwIiwib3JpZ2luIjoiMGx5bGU3M2EzMzIxMThjOWQzNzg4Iiwic3RibGsiOi0zNjAwLCJwbGFuIjowfQ.nj5sgxDzTp9tuLEVCxvrxXBJnqbCQH9MnoN6adcauBWPUwA7YpPG3AeijhCO5xMFyVP3Nw56oEURJybR879hGw
+
+TERITORI_DISTRIBUTOR_OWNER_MNEMONIC=1
+ETHEREUM_DISTRIBUTOR_OWNER_MNEMONIC=1
+PINATA_JWT=1
+AIRTABLE_API_KEY_DAPPS_STORE=patZYqro8XEjo8qXT.97183a7a2a21decc37eeafe8da99f669876df0df109bc908c6311f4302c1991a
```
Comment (omniwired) [inline .env:37]: please remove

File: networks.json [hunk 2]
```diff
@@ -282,20 +285,33 @@
     "accountExplorer": "https://goerli.etherscan.io/address/$address",
     "contractExplorer": "https://goerli.etherscan.io/address/$address",
     "testnet": true,
-    "backendEndpoint": "https://dapp-backend.testnet.teritori.com",
+    "backendEndpoint": "http://localhost:9090",
```
Comment (omniwired) [inline networks.json:288]: please change it back, before release

**Review by omniwired — CHANGES_REQUESTED** [review]

**Review by omniwired — APPROVED** [review]

File: package.json [hunk 3]
```diff
@@ -139,7 +139,7 @@
     "ts-node": "^10.9.1",
     "ts-proto": "^1.120.0",
     "ts-unused-exports": "^10.0.1",
-    "typescript": "~4.3.5",
+    "typescript": "~4.9.5",
```
Comment (omniwired) [inline package.json:142]: updated to fix some isuses

File: packages/screens/RiotGame/component/UnstakeModal.tsx [hunk 4]
```diff
@@ -43,8 +43,8 @@ export const UnstakeModal: React.FC<UnstakeModalProps> = ({
   const networkId = useSelectedNetworkId();
   const userInfo = useNSUserInfo(selectedWallet?.userId);
 
-  const startTime = squad?.start_time || 0;
-  const endTime = squad?.end_time || 0;
+  const startTime = squad?.startTime || 0;
+  const endTime = squad?.endTime || 0;
```
Comment (omniwired) [inline packages/screens/RiotGame/component/UnstakeModal.tsx:47]: 😸

File: go/pkg/p2e/seasons.go [hunk 5]
```diff
@@ -170,14 +170,30 @@ var THE_RIOT_SEASONS = []Season{
...
+var THE_RIOT_ETHEREUM_SEASONS = []Season{
+	// Season 1
+	{
+		ID:         "Season 1 - ETH",
+		GameID:     THE_RIOT_GAME_ID,
+		Denom:      "eth",
+		Decimals:   18,
+		TotalPrize: 1,
+		BossName:   "Philipp Rustov",
+		BossImage:  "",
```
Comment (omniwired) [inline go/pkg/p2e/seasons.go:194]: maybe we need to add an image here

File: packages/components/NetworkSelector.tsx [hunk 6]
```diff
@@ -0,0 +1,175 @@
...
+            .map((s) => s.trim())
+            .map(getNetwork)
+            .filter((n): n is NetworkInfo => !!n)
+            .filter((network) => {
+              return testnetsEnabled || !network.testnet;
+            })
+            .map((network, index) => {
+              const selectable =
+                !!selectableNetworks.find((sn) => sn.id === network.id) && // check that it's in the selectable list
+                selectedNetworkInfo?.id !== network.id && // check that it's not already selected
```
Comment (omniwired) [inline packages/components/NetworkSelector.tsx:142]: this looks hard to understand, are there any simplifications available?

File: packages/networks/ethereum/index.ts [hunk 7]
```diff
@@ -20,5 +20,13 @@ export const ethereumNetwork: NetworkInfo = {
   theGraphEndpoint:
     "https://api.studio.thegraph.com/query/40379/teritori-mainnet/v1",
   vaultContractAddress: "0x6251B3384c8eD53e2Cc38d34c1f26ffE8d461B94",
-  riotContractAddress: "0x8f8304ea566affeb96ad0ffb593bbebd8876d124",
+  riotContractAddressGen0: "0x8f8304ea566affeb96ad0ffb593bbebd8876d124",
+  riotSquadStakingContractAddress: "<must_fill>",
+  // Substreams
+  firehoseEndpoint: "<must_fill>",
```
Comment (omniwired) [inline packages/networks/ethereum/index.ts:26]: these will come from where?

File: packages/screens/Marketplace/SideCart.tsx [hunk 8]
```diff
@@ -340,6 +340,7 @@ const Footer: React.FC<{ items: any[] }> = ({ items }) => {
         } catch (e) {
           setToastError({
             title: "Error",
+            // @ts-expect-error
```
[the remark this answers is not in the document]
Comment (omniwired) [reply packages/screens/Marketplace/SideCart.tsx:343]: You need to update your typescript

File: go/pkg/marketplace/marketplace_service.go [hunk 9]
```diff
@@ -380,7 +368,7 @@ func (s *MarkteplaceService) NFTs(req *marketplacepb.NFTsRequest, srv marketplac
 		}
 		return nil
 
-	case *networks.CosmosNetwork:
+	case *networks.CosmosNetwork, *networks.EthereumNetwork:
```
[the remark this answers is not in the document]
Comment (hthieu1110) [reply go/pkg/marketplace/marketplace_service.go:371]: Yes, for this part of indexer, Cosmos and ETH behave the same. If I remember well, before using substreams to feed our indexer, we were using the graph which is 3rd indexer so there was another case for ETH

**Review by omniwired — DISMISSED** [review]

File: packages/screens/RiotGame/component/FightSection.tsx [hunk 10]
```diff
@@ -9,17 +9,99 @@ import { UnstakeModal } from "./UnstakeModal";
...
+  );
+  if (!metamaskProvider) {
+    throw Error("unable to get ethereum provider");
+  }
+
+  const nftInfos = await Promise.all(
+    squad.nfts.map(async (nft) => {
+      // If nft is bridged then use the original contract to query info
+      let nftContract = nft.contract;
+      let ethProvider: any = metamaskProvider;
```
[the remark this answers is not in the document]
Comment (hthieu1110) [reply packages/screens/RiotGame/component/FightSection.tsx:71]: Done

File: packages/screens/Marketplace/SideCart.tsx (hunk 8 above)
[the remark this answers is not in the document]
Comment (hthieu1110) [reply packages/screens/Marketplace/SideCart.tsx:343]: I cast e to any to remove // @ts-expect-error. 
Use of `any` is not ideal but in this case it allows us to remove `// @ts-expect-error`, which is slightly better IMHO

File: packages/networks/ethereum/index.ts (hunk 7 above)
Comment (hthieu1110) [reply packages/networks/ethereum/index.ts:30]: I will fill this once we test everything ok and once we have the contract mainnet ETH deployed

File: go/pkg/p2e/seasons.go (hunk 5 above)
Comment (hthieu1110) [reply go/pkg/p2e/seasons.go:194]: Yep it kind of place holder for now, I dont have yet the detail info about the first season of ETH p2e

**Review by omniwired — DISMISSED** [review]

**Review by WaDadidou — CHANGES_REQUESTED** [review]
Review WIP

**Pull request merged**